### ISSUE TITLE
GPU questions

### ISSUE THREAD

**Issue by MrOCR** [Calamari-OCR/calamari#37]
- Can I train using 2 GPUs?

- Does the new Nvidia RTX cards have any advantage over the 10xx series? Since RTX have `Tensor Cores`, would they substantially speed-up Calamari training?

The reason I am asking these questions because I am considering buying an RTX 2070, so is it worth it for Calamari training? 
https://www.phoronix.com/scan.php?page=article&item=nvidia-rtx2070-compute&num=5

**Comment by ChWick** [comment]
* we never tested training on several GPUs, however, Tensorflow does support multi-GPU training, thus this might work out of the box. Possibly, a high batch size is be required.
* The charts in your link suggest using 16-bit floats for the new RTX cards. However, Calamari uses 32-bit floats. We never tested 16-bit floats for training, and the effect on the accuracies. Thus, I cannot say if the RTX2070 is significantly faster than the older 10XX generations.

**Comment by MrOCR** [comment]
- In-order to train using multi-gpu, what would be the training command?

**Comment by ChWick** [comment]
* By default Tensorflow aquires all available Resources, therefore not additional commands should be required (check by nvidia-smi)
* Yes and no. Currently, calamari loads all data into the RAM before training, and streams it into tensorflow. If you have 5 TB of RAM this should work (*joking*). So, no, it will not work. However, one could consider to move some logic to the data streaming pipeline to load/unload/preprocess the data on-the-fly. Another problem is, that calamari automatically computes the codec by loading all text files and computing the set of all characters. This will take a long time for 5 TB of data, which is why the codec should be given by hand. Unfortunately, both are larger changes.
* Thus, you can currently only train on a small amount of data.

**Comment by MrOCR** [comment]
@ChWick Why didn't you use Mixed-Precision?
https://devblogs.nvidia.com/mixed-precision-training-deep-neural-networks/
https://arxiv.org/abs/1710.03740

**Comment by MrOCR** [comment]
@ChWick
So, Calamari training speed would scale by a factor of 2x using 2 GPUs?

**Comment by MrOCR** [comment]
@ChWick Thanks man for the `on-the-fly` branch! :sunglasses:

**Comment by MrOCR** [comment]
Note that you would need **Tensor Cores** to fully benefit from half-precision. Also note that the gtx 10x series are mostly optimized for fp32.

- https://github.com/NVIDIA/apex/issues/76
- https://en.wikipedia.org/wiki/List_of_Nvidia_graphics_processing_units#GeForce_20_series

![111](https://user-images.githubusercontent.com/33478216/48979375-25eb3900-f0cb-11e8-9330-ede6dedc5a77.jpg)

![222](https://user-images.githubusercontent.com/33478216/48979412-91350b00-f0cb-11e8-88aa-eec7bab4d7fc.jpg)
